src/dodal/devices/undulator.py

```diff
@@ -63,3 +95,58 @@ def __init__(
                 self.length = None
 
         super().__init__(name)
+
+    @AsyncStatus.wrap
+    async def set(self, value: float):
+        """
+        set the undulator gap to a given ENERGY
+
+        Args:
+            value: energy in keV
+        """
+        await asyncio.gather(
+            self._set_undulator_gap(value),
+        )
+
+    async def _set_undulator_gap(self, energy_kev: float) -> None:
+        access_level = await self.gap_access.get_value()
+        if access_level is UndulatorGapAccess.DISABLED and not TEST_MODE:
+            raise AccessError("Undulator gap access is disabled. Contact Control Room")
+        LOGGER.info(f"Setting undulator gap to {energy_kev:.2f} kev")
+        gap_to_match_dcm_energy = await self._gap_to_match_dcm_energy(energy_kev)
+
+        # Check if undulator gap is close enough to the value from the DCM
+        current_gap = await self.current_gap.get_value()
+        tolerance = await self.gap_discrepancy_tolerance_mm.get_value()
+        difference = abs(gap_to_match_dcm_energy - current_gap)
+        if difference > tolerance:
+            LOGGER.info(
+                f"Undulator gap mismatch. {difference:.3f}mm is outside tolerance.\
+                Moving gap to nominal value, {gap_to_match_dcm_energy:.3f}mm"
+            )
+            if not TEST_MODE:
+                # Only move if the gap is sufficiently different to the value from the
+                # DCM lookup table AND we're not in TEST_MODE
+                await self.gap_motor.set(
+                    gap_to_match_dcm_energy,
+                    timeout=STATUS_TIMEOUT_S,
+                )
+            else:
+                LOGGER.debug("In test mode, not moving ID gap")
+        else:
+            LOGGER.debug(
+                "Gap is already in the correct place for the new energy value "
+                f"{energy_kev}, no need to ask it to move"
+            )
+
+    async def _gap_to_match_dcm_energy(self, energy_kev: float) -> float:
```

Comment:
Should: As before, why are we referencing the `DCM` at all now? We wanted to decouple them and this can now run happily even if there is no DCM

Reply:
done